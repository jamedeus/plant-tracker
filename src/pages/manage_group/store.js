import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { groupSlice } from './groupSlice';
import plantDetailsProptypes from 'src/types/plantDetailsPropTypes';
import groupDetailsProptypes from 'src/types/groupDetailsPropTypes';

// Takes initial groupSlice state, returns redux store
function createReduxStore(preloadedState) {
    return configureStore({
        reducer: {
            group: groupSlice.reducer
        },
        preloadedState
    });
}

export function ReduxProvider({ children, initialState }) {
    const init = () => {
        // Return object with keys expected by groupSlice
        return {
            group: {
                groupDetails: initialState.group_details,
                plantDetails: initialState.plants
            },
        };
    };

    // Create redux store
    const store = useMemo(() => createReduxStore(
        init()
    ), [initialState]);

    return (
        <Provider store={store}>
            {children}
        </Provider>
    );
}

ReduxProvider.propTypes = {
    children: PropTypes.node,
    initialState: PropTypes.shape({
        group_details: groupDetailsProptypes.isRequired,
        plants: PropTypes.objectOf(plantDetailsProptypes).isRequired,
    }).isRequired
};
