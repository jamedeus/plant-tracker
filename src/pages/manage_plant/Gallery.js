import React, { useRef, useState, useMemo } from 'react';
import Lightbox from "yet-another-react-lightbox";
import Zoom from "yet-another-react-lightbox/plugins/zoom";
import Captions from "yet-another-react-lightbox/plugins/captions";
import Slideshow from "yet-another-react-lightbox/plugins/slideshow";
import Thumbnails from "yet-another-react-lightbox/plugins/thumbnails";
import Fullscreen from "yet-another-react-lightbox/plugins/fullscreen";
import "yet-another-react-lightbox/styles.css";
import "yet-another-react-lightbox/plugins/captions.css";
import "yet-another-react-lightbox/plugins/thumbnails.css";
import "src/css/gallery.css";
import { useSelector, useDispatch } from 'react-redux';
import { timestampToReadable } from 'src/timestampUtils';
import { photoGalleryOpened, photoGalleryIndexChanged } from './timelineSlice';
import { useIsBreakpointActive } from 'src/useBreakpoint';
import LoadingAnimation from 'src/components/LoadingAnimation';
import {
    XMarkIcon,
    PlayIcon,
    PauseIcon,
    ChevronLeftIcon,
    ChevronRightIcon,
    ArrowsPointingInIcon,
    ArrowsPointingOutIcon,
    MagnifyingGlassPlusIcon,
    MagnifyingGlassMinusIcon
} from '@heroicons/react/24/solid';

// Takes reference to element, returns true if within current viewport
const elementIsVisible = (element) => {
    const rect = element.getBoundingClientRect();
    return rect.top >= 0 && rect.bottom <= window.innerHeight;
};

const Gallery = () => {
    // Controls open state, set to true when timeline PhotoThumbnail clicked
    const open = useSelector((state) => state.timeline.photoGalleryOpen);
    // Controls initially visible photo (updates as carousel moves).
    // Used to open clicked photo + remember last-viewed photo when reopened.
    const index = useSelector((state) => state.timeline.photoGalleryIndex);
    // Controls slideshow delay (user-configurable in settings)
    const delay = useSelector((state) => state.settings.gallerySlideshowDelay);
    // Controls whether captions with photo date are visible
    const caption = useSelector((state) => state.settings.galleryShowPhotoDate);
    // Array of objects each representing 1 existing photo
    const photos = useSelector((state) => state.timeline.photos);
    const dispatch = useDispatch();

    // True if desktop layout, false if mobile
    const desktop = useIsBreakpointActive('md');

    // Shows progress animation if true
    const [slideshowRunning, setSlideshowRunning] = useState(false);

    // Track if slide has changed since opening (don't scroll timeline to
    // last-viewed photo unless slide has changed)
    const [slideHasChanged, setSlideHasChanged] = useState(false);

    // Get thumbnail + container heights for current layout
    // Pre-computing container height fixes animation stutter on mobile
    const thumbnailSize = desktop ? 100 : 80;
    const thumbnailContainerHeight = thumbnailSize + 32;

    // Build slides array, only update when redux photos state changes
    const slides = useMemo(() => photos.map(photo => ({
        src: photo.image,
        thumbnail: photo.thumbnail,
        description: timestampToReadable(photo.timestamp).split('-')[1],
        imageFit: 'contain',
        key: photo.key
    })), [photos]);

    return (
        <Lightbox
            open={open}
            on={{
                // Reset slideHasChanged when opened
                entered: () => {
                    setSlideHasChanged(false);
                },

                // Scroll timeline to last-viewed photo before closing
                exiting: () => {
                    if (slideHasChanged) {
                        // Get reference to last-viewed photo thumbnail
                        const currentSlide = slides[index];
                        const thumbnail = document.querySelector(
                            `[data-timeline-thumbnail="${currentSlide.thumbnail}"]`
                        );
                        // Only scroll if thumbnail is outside viewport
                        if (!elementIsVisible(thumbnail)) {
                            thumbnail.scrollIntoView(
                                { behavior: "smooth", block: "start" }
                            );
                        }
                    }
                },

                // Close gallery
                exited: () => dispatch(photoGalleryOpened({ open: false })),

                // Track current slide each time slide changes
                view: ({ index: currentIndex }) => {
                    // Don't set slideHasChanged when lightbox first opens
                    if (currentIndex !== index) {
                        setSlideHasChanged(true);
                    }
                    dispatch(photoGalleryIndexChanged({ index: currentIndex }));
                },

                // Start progress animation when slideshow starts
                slideshowStart: () => {
                    setSlideshowRunning(true);
                },

                // Stop progress animation when slideshow stops
                slideshowStop: () => {
                    setSlideshowRunning(false);
                },
            }}
            render={{
                // Custom icons (match icons used in rest of the app)
                iconPrev: () => <ChevronLeftIcon className="size-8" />,
                iconNext: () => <ChevronRightIcon className="size-8" />,
                iconZoomIn: () => <MagnifyingGlassPlusIcon className="size-8" />,
                iconZoomOut: () => <MagnifyingGlassMinusIcon className="size-8" />,
                iconSlideshowPlay: () => <PlayIcon className="size-8" />,
                iconSlideshowPause: () => <PauseIcon className="size-8" />,
                iconClose: () => <XMarkIcon className="size-8" />,
                iconExitFullscreen: () => (<ArrowsPointingInIcon className="size-8" />),
                iconEnterFullscreen: () => <ArrowsPointingOutIcon className="size-8" />,
                // Match loading animation used in rest of app
                iconLoading: () => <LoadingAnimation />,
                // Render progress bar if slideshow is running
                // Key changes on each slide (remount, start animation over)
                controls: () => {
                    if (slideshowRunning) {
                        return (
                            <div
                                key={index}
                                className="slideshow_progress_bar"
                                style={{
                                    "--slideshow-delay": `${delay}ms`
                                }}
                            />
                        );
                    }
                }
            }}
            controller={{
                closeOnPullDown: true
            }}
            plugins={[
                Zoom,
                ...(caption ? [Captions] : []),
                Slideshow,
                Thumbnails,
                Fullscreen
            ]}
            toolbar={{
                buttons: [
                    "slideshow",
                    "fullscreen",
                    "close"
                ]
            }}
            index={index}
            slides={slides}
            thumbnails={{
                width: thumbnailSize,
                height: thumbnailSize,
                border: 0,
                padding: 0,
                imageFit: 'cover',
                vignette: true,
            }}
            slideshow={{
                delay: delay
            }}
            zoom={{
                // Limit zoom to 50% on mobile
                maxZoomPixelRatio: desktop ? 1 : 0.5
            }}
            // Set class that causes thumbnails to shrink when slideshow starts
            className={slideshowRunning ? 'slideshow-running' : null}
            // Set pre-computed height var used in thumbnail shrink transition
            // Fixes stutter on mobile caused by using calc() for initial height
            styles={{
                thumbnailsContainer: {
                    "--thumbnail-container-height": `${thumbnailContainerHeight}px`
                }
            }}
        />
    );
};

export default Gallery;
