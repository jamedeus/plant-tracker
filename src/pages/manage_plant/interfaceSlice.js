import { createSlice } from '@reduxjs/toolkit';

// Redux slice to store user interface state
export const interfaceSlice = createSlice({
    name: 'interface',
    initialState: {
        // Navbar title drawer open state
        titleDrawerOpen: false,
        // Fullscreen scanner used to change QR code open state
        changeQrScannerOpen: false,
        // Settings side menu open state
        settingsMenuOpen: false,
        // Fullscreen lightbox gallery open state
        photoGalleryOpen: false,
        // Currently visible photo in lightbox gallery (index of photos state)
        photoGalleryIndex: 0,
        // Can select events/photos/notes to delete by clicking in timeline if true
        deleteMode: false,
        // Tracks events selected by clicking in timeline
        // Uses same format as /bulk_delete_plant_events payload
        selectedEvents: {
            water: [],
            fertilize: [],
            prune: [],
            repot: []
        },
        // Tracks photos selected by clicking in timeline (array of photo keys)
        selectedPhotos: [],
        // Tracks notes selected by clicking in timeline (array of note timestamps)
        selectedNotes: []
    },
    reducers: {
        // Takes true or false
        titleDrawerOpened(state, action) {
            state.titleDrawerOpen = action.payload;
        },

        // Takes true or false
        changeQrScannerOpened(state, action) {
            state.changeQrScannerOpen = action.payload;
            state.titleDrawerOpen = false;
        },

        // Takes true or false
        settingsMenuOpened(state, action) {
            state.settingsMenuOpen = action.payload;
            // Close dropdown (visible through transparent settings menu)
            document.activeElement.blur();
        },

        // Takes { open: <bool> }
        photoGalleryOpened(state, action) {
            state.photoGalleryOpen = action.payload.open;
        },

        // Takes { index: <int> }
        photoGalleryIndexChanged(state, action) {
            state.photoGalleryIndex = action.payload.index;
        },

        // Takes { editing: <bool> }
        deleteModeChanged(state, action) {
            state.deleteMode = action.payload.editing;
            // Clear selections if no longer deleting
            if (!action.payload.editing) {
                state.selectedEvents = {
                    water: [],
                    fertilize: [],
                    prune: [],
                    repot: []
                };
                state.selectedPhotos = [];
                state.selectedNotes = [];
            }
        },

        // Takes { type: <eventType, timestamps: <array>, selected: <bool> }
        // Timestamps must be ISO format (UTC, not user timezone)
        eventSelected(state, action) {
            // Add timestamps to selectedEvents if selected bool is true
            if (action.payload.selected){
                state.selectedEvents[action.payload.type] = [
                    ...state.selectedEvents[action.payload.type],
                    ...action.payload.timestamps
                ];
            // Remove timestamps from selectedEvents if selected bool is false
            } else {
                state.selectedEvents[action.payload.type] =
                    state.selectedEvents[action.payload.type].filter(
                        timestamp => !action.payload.timestamps.includes(timestamp)
                    );
            }
        },

        // Takes { key: <int>, selected: <bool> }
        photoSelected(state, action) {
            if (action.payload.selected) {
                state.selectedPhotos.push(action.payload.key);
            } else {
                state.selectedPhotos = state.selectedPhotos.filter(
                    key => key !== action.payload.key
                );
            }
        },

        // Takes { timestamp: <string>, selected: <bool> }
        noteSelected(state, action) {
            if (action.payload.selected) {
                state.selectedNotes.push(action.payload.timestamp);
            } else {
                state.selectedNotes = state.selectedNotes.filter(
                    timestamp => timestamp !== action.payload.timestamp
                );
            }
        },
    }
});

export const {
    titleDrawerOpened,
    changeQrScannerOpened,
    settingsMenuOpened,
    photoGalleryOpened,
    photoGalleryIndexChanged,
    deleteModeChanged,
    eventSelected,
    photoSelected,
    noteSelected,
} = interfaceSlice.actions;
