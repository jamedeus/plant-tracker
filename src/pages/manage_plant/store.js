import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { parseDomContext } from 'src/util';
import { timestampToDateString } from 'src/timestampUtils';
import { plantSlice } from './plantSlice';
import { timelineSlice } from './timelineSlice';
import { settingsSlice } from './settingsSlice';
import { loadUserSettings } from './Settings';
import { useIsBreakpointActive } from 'src/useBreakpoint';

// Takes object with event type keys, array of timestamps as value.
// Converts to object with YYYY-MM-DD keys, each containing an object with
// "events", "notes", and "photos" keys used to populate timeline.
const formatEvents = (events) => {
    return Object.entries(events).reduce(
        (acc, [eventType, eventDates]) => {
            eventDates.forEach(date => {
                const dateKey = timestampToDateString(date);
                // Add new date key unless it already exists
                if (!acc[dateKey]) {
                    acc[dateKey] = {events: [], notes: [], photos: []};
                }
                // Add event to date key unless same type already exists
                if (!acc[dateKey]['events'].includes(eventType)) {
                    acc[dateKey]['events'].push(eventType);
                }
            });
            return acc;
        },
        {}
    );
};

// Takes events, notes, and photos context objects from django backend, returns
// timelineDays state used by Timeline component (YYYY-MM-DD keys containing
// objects with events, notes, and photos keys).
export const buildTimelineDays = (events, notes, photos) => {
    // Convert to object with YYYY-MM-DD keys
    const timelineDays = formatEvents(events);

    // Add objects from photos context to photos key under correct dateKey
    photos.sort((a, b) => {
        return a.timestamp.localeCompare(b.timestamp);
    }).reverse();
    photos.forEach((photo) => {
        const dateKey = timestampToDateString(photo.timestamp);
        if (!timelineDays[dateKey]) {
            timelineDays[dateKey] = {events: [], notes: [], photos: []};
        }
        timelineDays[dateKey].photos.push(photo);
    });

    // Add objects from notes context to notes key under correct dateKey
    notes.forEach((note) => {
        const dateKey = timestampToDateString(note.timestamp);
        if (!timelineDays[dateKey]) {
            timelineDays[dateKey] = {events: [], notes: [], photos: []};
        }
        timelineDays[dateKey].notes.push(note);
    });

    return timelineDays;
};

// Takes timelineDays state, flattens and returns calendarDays state used by
// EventCalendar component (YYYY-MM-DD keys containing array of event strings).
export const buildCalendarDays = (timelineDays) => {
    const calendarDays = {};
    Object.keys(timelineDays).forEach(dateKey => {
        if (timelineDays[dateKey].events.length) {
            calendarDays[dateKey] = timelineDays[dateKey].events;
        }
    });
    return calendarDays;
};

// Takes timelineDays state, returns navigationOptions state used to populate
// timeline QuickNavigation options (YYYY keys containing array of MM strings).
export const buildNavigationOptions = (timelineDays) => {
    const navigationOptions = {};
    Object.keys(timelineDays).forEach(dateString => {
        const [year, month] = dateString.split('-');
        if (!navigationOptions[year]) {
            navigationOptions[year] = [];
        }
        if (!navigationOptions[year].includes(month)) {
            navigationOptions[year].push(month);
        }
    });
    // Sort months in each year chronologically (timelineDays is not sorted)
    Object.keys(navigationOptions).forEach(year =>
        navigationOptions[year].sort().reverse()
    );
    return navigationOptions;
};

// Takes initial plantSlice and timelineSlice states, returns redux store
function createReduxStore(preloadedState) {
    return configureStore({
        reducer: {
            plant: plantSlice.reducer,
            timeline: timelineSlice.reducer,
            settings: settingsSlice.reducer
        },
        preloadedState
    });
}

export function ReduxProvider({ children }) {
    // Get layout string used to look up default settings for current breakpoint
    const layout = useIsBreakpointActive("md") ? 'desktop' : 'mobile';

    // Parses django context elements containing events, photos, and notes
    // Merges and returns values for all initialState keys in timelineSlice
    const init = () => {
        // Parse django context objects
        const plantDetails = parseDomContext("plant_details");
        const groupOptions = parseDomContext("group_options");
        const eventsByType = parseDomContext("events");
        const photos = parseDomContext('photos');
        const notes = parseDomContext('notes');
        const defaultPhoto = parseDomContext('default_photo');
        const photoGalleryIndex = 0;
        const photoGalleryOpen = false;

        // Build state objects
        const timelineDays = buildTimelineDays(eventsByType, notes, photos);
        const calendarDays = buildCalendarDays(timelineDays);
        const navigationOptions = buildNavigationOptions(timelineDays);

        // Return object with keys expected by plantSlice and timelineSlice
        return {
            plant: {
                plantDetails,
                groupOptions
            },
            timeline: {
                eventsByType,
                calendarDays,
                timelineDays,
                photos,
                navigationOptions,
                defaultPhoto,
                photoGalleryIndex,
                photoGalleryOpen
            },
            settings: loadUserSettings(layout)
        };
    };

    // Create redux store with initial state built from django context items
    const store = useMemo(() => createReduxStore(
        init()
    ), []);

    return (
        <Provider store={store}>
            {children}
        </Provider>
    );
}

ReduxProvider.propTypes = {
    children: PropTypes.node
};
