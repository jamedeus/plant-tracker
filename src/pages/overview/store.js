import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { overviewSlice } from './overviewSlice';

// Takes initial plantSlice and timelineSlice states, returns redux store
function createReduxStore(preloadedState) {
    return configureStore({
        reducer: {
            overview: overviewSlice.reducer
        },
        preloadedState
    });
}

export function ReduxProvider({ children, initialState }) {
    const init = () => {
        // Check URL to determine if viewing main overview or archive overview
        const archivedOverview = window.location.pathname === '/archived';

        // Return object with keys expected by overviewSlice
        return {
            overview: {
                plants: initialState.plants,
                groups: initialState.groups,
                archivedOverview: archivedOverview,
                showArchive: initialState.show_archive,
                title: initialState.title
            },
        };
    };

    // Create redux store
    const store = useMemo(() => createReduxStore(
        init()
    ), [initialState]);

    return (
        <Provider store={store}>
            {children}
        </Provider>
    );
}

ReduxProvider.propTypes = {
    children: PropTypes.node,
    initialState: PropTypes.shape({
        plants: PropTypes.object.isRequired,
        groups: PropTypes.object.isRequired,
        show_archive: PropTypes.bool.isRequired,
        title: PropTypes.string.isRequired
    }).isRequired
};
