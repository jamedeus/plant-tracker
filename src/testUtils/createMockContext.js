// Creates DOM elements containing JSON (mocks django json_script filter)
// Takes context id (arg passed to parseDomContext in App.js) and json contents
const createMockContext = (id, json) => {
    const mockContext = document.createElement('div');
    mockContext.id = id;
    mockContext.textContent = JSON.stringify(json);
    document.body.appendChild(mockContext);
};

export default createMockContext;
