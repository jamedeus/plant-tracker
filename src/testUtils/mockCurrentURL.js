// Takes URL and optional pathname, mocks window.location
const mockCurrentURL = (url, pathname) => {
    Object.defineProperty(window, 'location', {
        configurable: true,
        value: {
            ...window.location,
            href: url,
            pathname: pathname,
            origin: url.replace(pathname, ''),
            assign: jest.fn()
        }
    });
};

export default mockCurrentURL;
