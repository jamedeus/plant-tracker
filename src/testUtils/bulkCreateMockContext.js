import createMockContext from './createMockContext';

// Takes nested mockContext JSON (simulates backend render_react_app state arg)
// Iterates and creates DOM elements with ID of each key containing contents of
// each key (simulates django json_script filter used in index.html)
const bulkCreateMockContext = (json) => {
    Object.entries(json).forEach(([key, value]) => {
        createMockContext(key, value);
    });
};

export default bulkCreateMockContext;
