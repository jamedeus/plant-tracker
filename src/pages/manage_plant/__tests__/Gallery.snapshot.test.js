import createMockContext from 'src/testUtils/createMockContext';
import mockCurrentURL from 'src/testUtils/mockCurrentURL';
import bulkCreateMockContext from 'src/testUtils/bulkCreateMockContext';
import { mockContext, mockphotos } from './mockContext';
import { ReduxProvider } from '../store';
import { useDispatch } from 'react-redux';
import Gallery from '../Gallery';
import { photoGalleryOpened } from '../interfaceSlice';

const TestComponent = () => {
    const dispatch = useDispatch();

    return (
        <>
            <button onClick={() => dispatch(photoGalleryOpened({open: true}))}>
                Open Gallery
            </button>
            <Gallery />
        </>
    );
};

describe('Gallery', () => {
    beforeAll(() => {
        // Create mock state objects
        bulkCreateMockContext(mockContext);
        createMockContext('user_accounts_enabled', true);
        // Override photos state with mock containing more photos
        createMockContext('photos', mockphotos);
        // Mock window.location (querystring parsed when page loads)
        mockCurrentURL('https://plants.lan/manage/e1393cfd-0133-443a-97b1-06bb5bd3fcca');
    });

    it('matches snapshot for desktop layout', async () => {
        // Set width greater than tailwind md breakpoint, render gallery
        window.innerWidth = 800;
        const user = userEvent.setup();
        const { container, getByRole } = render(
            <ReduxProvider>
                <TestComponent />
            </ReduxProvider>
        );

        // Open gallery, confirm matches snapshot
        await user.click(getByRole('button', {name: 'Open Gallery'}));
        expect(container).toMatchSnapshot();
    });

    it('matches snapshot for desktop layout', async () => {
        // Set width less than tailwind md breakpoint, render gallery
        window.innerWidth = 600;
        const user = userEvent.setup();
        const { container, getByRole } = render(
            <ReduxProvider>
                <TestComponent />
            </ReduxProvider>
        );

        // Open gallery, confirm matches snapshot
        await user.click(getByRole('button', {name: 'Open Gallery'}));
        expect(container).toMatchSnapshot();
    });
});

