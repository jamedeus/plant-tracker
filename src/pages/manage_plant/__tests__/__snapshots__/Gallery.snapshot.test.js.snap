// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery matches snapshot for desktop layout 1`] = `
<div
  aria-hidden="true"
  inert=""
>
  <button>
    Open Gallery
  </button>
</div>
`;

exports[`Gallery matches snapshot for desktop layout 2`] = `
<div
  aria-hidden="true"
  inert=""
>
  <button>
    Open Gallery
  </button>
</div>
`;
